### ISSUE TITLE
client-go: keep API versions longer in the client library than in the apiserver

### ISSUE THREAD

**Issue by sttts** [kubernetes/kubernetes#56985]
One pretty uniform feedback from client-go users @ KubeCon was that they struggle with supporting old and new Kubernetes clusters from one code-base of their operators. Keeping old API versions around longer than we serve them in kube-apiserver might help them. So they can move on to the new client-go, but have an abstraction logic in their operator to fall-back to the old API group versions.

Keeping API group versions around is cheap: we only have to keep them in k8s.io/api, but can remove the internal types and conversions.

**Comment by caesarxuchao** [comment]
Are the complaints mostly about alpha API? Beta API is supposed to be kept around for 6 months after deprecation. Or is 6 months still not long enough?

**Comment by ncdc** [comment]
I would imagine there are a good number of clusters still running 1.3, 1.4, 1.5, 1.6, etc.

**Comment by sttts** [comment]
Compare https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#alpha-beta-and-stable-versions:

If 1.9 tomorrow has abc/v1, then abc/v1beta1 can go away in 1.10 in 3 months. That's not a lot.

**Comment by caesarxuchao** [comment]
Another doc says 6 months: https://kubernetes.io/docs/reference/deprecation-policy/

`older API versions must be supported after their announced deprecation for a duration of no less than:...Beta: 6 months or 2 releases (whichever is longer)`

Which one is correct? @kubernetes/sig-architecture-bugs @smarterclayton @bgrant0607 ?

**Comment by ncdc** [comment]
The proposal is that there could potentially be a difference between how long an apiserver supports an API and how long client-go supports the same API (post deprecation)... right?

**Comment by sttts** [comment]
> right?

yes

**Comment by caesarxuchao** [comment]
@ncdc Yes.

My previous comment just pointed out that sttts and I found a discrepancy in our docs, one claiming the deprecation of beta API take 3 months, the other one claiming it's 6 months.

**Comment by bgrant0607** [comment]
Where there is a discrepancy,  https://kubernetes.io/docs/reference/deprecation-policy/ is authoritative.

**Comment by bgrant0607** [comment]
In this case, I believe there isn't actually a discrepancy. api_changes.md is stating a technical requirement, not the deprecation policy.

**Comment by ncdc** [comment]
That and/or the ability to take an older client-go version and generate updated clientsets for it to pull in new api groups/versions/types.

**Comment by caesarxuchao** [comment]
> /cc @LiliC @brancz – any comments on your needs?

Do you need to support clusters span more than 3 releases? For apiservers whose versions differ by more than 3 versions, there could be incompatible changes other than the kubernetes APIs.

**Comment by brancz** [comment]
Ideally we would like to enable users to use the Prometheus Operator with any supported Kubernetes version, meaning those that still actively get security updates (by quick googling I couldn't find documentation on which versions this would mean concretely today). That would mean that we would like to have all APIs/versions in client-go down to the APIs available in those releases.

My guess is that for us this is going to get less complicated with 1.9 as the majority of APIs used are moving to stable, but I can see this to continue to be a problem with new/other non-stable APIs.

**Comment by LiliC** [comment]
I agree with @brancz, but I can see the documentation of this to become easily outdated. I would propose to just have a fixed number of releases to support, that way it's easy to drop support docs and add new support docs for new releases.

**Comment by brancz** [comment]
/reopen

I think this is still relevant.
